source/module_hsolver/diago_elpa_native.cpp

```diff
@@ -0,0 +1,150 @@
+#include "diago_elpa_native.h"
+
+#include "module_base/blacs_connector.h"
+#include "module_base/global_variable.h"
+#include "module_base/lapack_connector.h"
+#include "module_base/scalapack_connector.h"
+#include "module_base/timer.h"
+#include "module_base/tool_quit.h"
+#include "module_hsolver/genelpa/elpa_new.h"
+#include "omp.h"
+
+namespace hsolver
+{
+#ifdef __MPI
+template <typename T>
+MPI_Comm DiagoElpaNative<T>::setmpicomm()
+{
+    if (this->elpa_num_thread == -1)
+    {
+        return MPI_COMM_WORLD;
+    }
+    else
+    {
+        int _num = 0;
+        MPI_Comm_size(MPI_COMM_WORLD, &_num);
+        if (elpa_num_thread > _num || elpa_num_thread <= 0)
+        {
+            return MPI_COMM_WORLD;
+        }
+        else
+        {
+            lastmpinum++;
+            int* _ranks = new int[elpa_num_thread];
+            for (int i = 0; i < elpa_num_thread; i++)
+            {
+                _ranks[i] = (lastmpinum + i) % _num;
+            }
+            MPI_Group _tempgroup, _oldgroup;
+            MPI_Comm_group(MPI_COMM_WORLD, &_oldgroup);
+            MPI_Group_incl(_oldgroup, elpa_num_thread, _ranks, &_tempgroup);
+            MPI_Comm _new_comm;
+            MPI_Comm_create(MPI_COMM_WORLD, _tempgroup, &_new_comm);
+            delete[] _ranks;
+            return _new_comm;
+        }
+    }
+}
+#endif
+
+#ifdef __MPI
+template <typename T>
+void DiagoElpaNative<T>::diag_pool(hamilt::MatrixBlock<T>& h_mat,
+                                   hamilt::MatrixBlock<T>& s_mat,
+                                   psi::Psi<T>& psi,
+                                   Real* eigenvalue_in,
+                                   MPI_Comm& comm)
+{
+    std::vector<Real> eigen(GlobalV::NLOCAL, 0.0);
+
+    ModuleBase::timer::tick("DiagoElpaNative", "elpa_solve");
+
+    int nev = GlobalV::NBANDS;
```

Comment:
avoid using GlobalV::NBANDS, maybe the variable can be passed into this function?

Reply:
If we want to modify this interface, we need to modify all the interfaces for generalized eigenvalues ​​at the same time, because it involves changes to the base class interface. I think it would be best to deal with this in other PRs in the future.